### PR TITLE
feat: create button group widget

### LINKED ISSUE #4889: [Feature] Button Group Widget
## Summary

A widget to create a group of buttons and menus inside a single container.

Ex:
Screenshot 2021-06-03 at 12 15 01 PM
 
## Requirements

Property | Values | Default | Definition
-- | -- | -- | --
orientation | 'horizontal'\|&nbsp;'vertical' | 'horizontal' | The group orientation (layout flow direction).
variant | 'contained'\|&nbsp;'outlined'\|&nbsp;'text' | 'outlined' | The variant to use.

## Design

https://www.figma.com/file/XT0We8ms6XC3iD73lTF2Zr/Button-Group-Widget?node-id=2%3A262

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Button Group widget 

Fixes #4889

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x]  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/new-btn-group-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.81 **(-0.1)** | 36.48 **(-0.14)** | 33.48 **(-0.12)** | 55.36 **(-0.09)**
 :sparkles: :new: | **app/client/src/components/propertyControls/ButtonListControl.tsx** | **23.66** | **29.41** | **7.69** | **25.61**
 :green_circle: | app/client/src/components/propertyControls/index.ts | 85.29 **(0.44)** | 100 **(0)** | 0 **(0)** | 84.85 **(0.47)**
 :red_circle: | app/client/src/icons/WidgetIcons.tsx | 62.67 **(-0.34)** | 100 **(0)** | 20 **(-0.59)** | 62.67 **(-0.34)**
 :sparkles: :new: | **app/client/src/widgets/ButtonGroupWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/ButtonGroupWidget/component/index.tsx** | **26.25** | **11.38** | **5.26** | **28.57**
 :sparkles: :new: | **app/client/src/widgets/ButtonGroupWidget/widget/index.tsx** | **59.26** | **0** | **33.33** | **60**</details>

### CODE REVIEW

File: app/client/src/components/propertyControls/ButtonListControl.tsx [hunk 1]
```diff
@@ -0,0 +1,250 @@
+import React, { useCallback } from "react";
```
Comment (jsartisan) [inline app/client/src/components/propertyControls/ButtonListControl.tsx:1]: Quick question: which control is this in UI?

Comment (jsartisan) [reply app/client/src/components/propertyControls/ButtonListControl.tsx:1]: Don't we already have control for these things? can't we reuse what we have for menu widget/tabs/table. or do we need to make new control for every new widget?

**Comment by jsartisan** [comment]
@techbhavin code-wise it looks good me. just had few doubts regarding the new control file.

**Review by jsartisan — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin Tested this feature and please find the issues below,

1.When adding more buttons with or without icon the alignment of button name and icon is not unique 

Screenshot 2021-09-22 at 1 23 19 PM

2.When any of the button type is menu and which has sub menu's - Then the sub menu's are displayed outside the widget in Horizontal orientation but in vertical orientation the sub menu's are displayed inside widget 

https://www.loom.com/share/97ea70f772ae4ed398a4d51c38a778d9

3. Button Style doesn't have primary option and by default for all the button style - The default option is secondary - white colour and which actually should be light brown colour - similar to button widget 

https://www.loom.com/share/d92f31d314d94893bac45de3295f6db7

4. Enabling disable toggle icon - Still the widget is not disabled 

https://www.loom.com/share/300fc19a6b8748a1bab506c673129b6e

5.Suggestion to have scroll functionality - if there are more button then the button name gets overlapped in horizontal orientation - If we implement scroll then when are more button then we can avoid overlaps 

Screenshot 2021-09-22 at 2 03 48 PM

**Comment by aswathkk** [comment]
@techbhavin 
We are moving away from buttonStyle to buttonColor to avoid confusion.
Also, we are changing buttonVariant labels from solid, outline, ghost to primary, secondary and tertiary. You can see the discussions regarding this [here](https://github.com/appsmithorg/appsmith/issues/7046)
And [here](https://github.com/appsmithorg/appsmith/pull/7166)'s the PR which does the above change for Button, IconButton and MenuButton widgets.

Cc: @somangshu

**Comment by somangshu** [comment]
@techbhavin @YogeshJayaseelan these are the latest changes we need to be consistent with ☝️, as soon as #7166 is merged, we need to make the same changes here.

**Review by jsartisan — DISMISSED** [review]

**Comment by jsartisan** [comment]
@somangshu We have removed the border-radius and shadow in the default config just to maintain consistency.  Let me or bhavin know if you have any feedback on that.

**Review by jsartisan — APPROVED** [review]

**Comment by somangshu** [comment]
> @somangshu We have removed the border-radius and shadow in the default config just to maintain consistency. Let me or bhavin know if you have any feedback on that.

@jsartisan didnt exactly understand this, What is default config, Did you mean general section of the property pane. Please take that call in favour of consistency.

Another thing is if the feature is covered by tests, we need to make sure of that. And docs as well. 

cc @techbhavin

**Comment by techbhavin** [comment]
Document is update is [Here](https://github.com/appsmithorg/appsmith-docs/pull/205)

cc: @somangshu

**Review by jsartisan — DISMISSED** [review]

**Comment by somangshu** [comment]
@momcilo-appsmith the challenge we face in a button group is that the alignment looks incorrect when the icon and label are center aligned and the labels have different length. I found that many other frameworks follow the same approach (attached an example above).

if this looks alright to you guys, please let us know, We would like to go ahead and merge this!

**Comment by YogeshJayaseelan** [comment]
@techbhavin 

For Point 1 - We will raise a new issue 

and the remaining part looks good and we can proceed further

**Comment by somangshu** [comment]
@techbhavin please check if fixes related to button / variant etc is merged and this branch is updated, We need to make sure that it does not create any problems in the widget.

**Comment by somangshu** [comment]
> @techbhavin please check if fixes related to button / variant etc is merged and this branch is updated, We need to make sure that it does not create any problems in the widget.

@jsartisan can you verify this

**Review by jsartisan — APPROVED** [review]

**Comment by techbhavin** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1447514766

**Comment by techbhavin** [comment]
https://github.com/appsmithorg/appsmith/runs/4174651660

**Pull request merged**